### PR TITLE
Fix stats to metrics implementation.

### PR DESCRIPTION
**Pull request by bogdandrutu**
In the metrics data model the "Point" does not represent individual points, it represents an individual point in time in the TimeSeries and in case of cumulative metrics this is the cumulative value from the start of the TimeSeries.

### CODE REVIEW

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java [hunk 1]
```diff
@@ -119,6 +125,23 @@ private CumulativeMutableViewData(View view, Timestamp start, MetricMap metricMa
       }
     }
 
+    @Override
+    Metric toMetric(Timestamp now, StatsCollectionState state) {
+      if (state == StatsCollectionState.DISABLED) {
+        return null;
+      }
+      MetricDescriptor.Type type = metricDescriptor.getType();
+      List<TimeSeriesCumulative> timeSeriesCumulatives = new ArrayList<TimeSeriesCumulative>();
```
Comment (songy23) [inline impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: This list should be created based on `type`. If `type` is `GAUGE_DOUBLE` or `GAUGE_INT64`, this should be `TimeSeriesGauge`. See https://github.com/census-instrumentation/opencensus-java/pull/1368/files#diff-05a76803285c4ebb76d41d8822baacbfL117.

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java [hunk 2]
```diff
@@ -119,6 +125,23 @@ private CumulativeMutableViewData(View view, Timestamp start, MetricMap metricMa
...
+      for (Entry<List<TagValue>, MutableAggregation> entry : tagValueAggregationMap.entrySet()) {
+        List<LabelValue> labelValues = MetricUtils.tagValuesToLabelValues(entry.getKey());
+        Point point = MetricUtils.mutableAggregationToPoint(entry.getValue(), now, type);
+        timeSeriesCumulatives.add(
+            TimeSeriesCumulative.create(labelValues, Collections.singletonList(point), start));
```
Comment (songy23) [inline impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: If each `TimeSeries` would always have only one `Point`, why do we need to define `Point`s as a list in the proto?

```proto
// A collection of data points that describes the time-varying values
// of a metric.
message TimeSeries {
  // Must be present for cumulative metrics. The time when the cumulative value
  // was reset to zero. The cumulative value is over the time interval
  // [start_timestamp, timestamp]. If not specified, the backend can use the
  // previous recorded value.
  google.protobuf.Timestamp start_timestamp = 1;

  // The set of label values that uniquely identify this timeseries. Applies to
  // all points. The order of label values must match that of label keys in the
  // metric descriptor.
  repeated LabelValue label_values = 2;

  // The data points of this timeseries. Point.value type MUST match the
  // MetricDescriptor.type.
  repeated Point points = 3;
}
```

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java [hunk 3]
```diff
@@ -31,19 +31,26 @@
...
 import io.opencensus.stats.ViewData;
 import io.opencensus.tags.TagContext;
 import io.opencensus.tags.TagValue;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import javax.annotation.Nullable;
```
Comment (songy23) [inline impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: Please fix the import. You'll need to use full qualifier for `javax.annotation.Nullable`.

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java (hunk 1 above)
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: Done.

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java (hunk 2 above)
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: Multiple reasons why:
1) Same proto can be used for getting data. So in the storage you have more points.
2) If we produce a point each second but export every 10 seconds (not implemented this way right now).

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java (hunk 3 above)
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: done.

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java [hunk 4]
```diff
@@ -119,6 +127,35 @@ private CumulativeMutableViewData(View view, Timestamp start, MetricMap metricMa
...
+    @Override
+    Metric toMetric(Timestamp now, StatsCollectionState state) {
+      if (state == StatsCollectionState.DISABLED) {
+        return null;
+      }
+      // TODO(bdrutu): Refactor this after TimeSeriesGauge and TimeSeriesCumulative are combined.
+      Type type = metricDescriptor.getType();
+      if (type == Type.GAUGE_INT64 || type == Type.GAUGE_DOUBLE) {
+        List<TimeSeriesGauge> timeSeriesGauges = new ArrayList<TimeSeriesGauge>();
+        for (Entry<List<TagValue>, MutableAggregation> entry : tagValueAggregationMap.entrySet()) {
```
Comment (songy23) [inline impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: `TagValue` -> `/*@Nullable*/ TagValue`

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java [hunk 5]
```diff
@@ -119,6 +127,35 @@ private CumulativeMutableViewData(View view, Timestamp start, MetricMap metricMa
...
+          List<LabelValue> labelValues = MetricUtils.tagValuesToLabelValues(entry.getKey());
+          Point point = MetricUtils.mutableAggregationToPoint(entry.getValue(), now, type);
+          timeSeriesGauges.add(
+              TimeSeriesGauge.create(labelValues, Collections.singletonList(point)));
+        }
+        return Metric.create(metricDescriptor, TimeSeriesGaugeList.create(timeSeriesGauges));
+      } else {
+        List<TimeSeriesCumulative> timeSeriesCumulatives = new ArrayList<TimeSeriesCumulative>();
+        for (Entry<List<TagValue>, MutableAggregation> entry : tagValueAggregationMap.entrySet()) {
```
Comment (songy23) [inline impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: ditto

File: impl_core/src/main/java/io/opencensus/implcore/stats/MeasureToViewMap.java [hunk 6]
```diff
@@ -161,16 +154,24 @@ synchronized void record(TagContext tags, MeasureMapInternal stats, Timestamp ti
...
     }
   }
 
-  synchronized List<Metric> getMetrics() {
-    return metricMap.toMetrics();
+  synchronized List<Metric> getMetrics(Clock clock, StatsCollectionState state) {
+    List<Metric> metrics = new ArrayList<Metric>();
+    Timestamp now = clock.now();
+    for (Entry<String, MutableViewData> entry : mutableMap.entries()) {
+      Metric metric = entry.getValue().toMetric(now, state);
```
Comment (songy23) [inline impl_core/src/main/java/io/opencensus/implcore/stats/MeasureToViewMap.java]: `Metric` -> `/*@Nullable*/ Metric`

File: impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java (hunk 2 above)
Comment (songy23) [reply impl_core/src/main/java/io/opencensus/implcore/stats/MutableViewData.java]: SGTM

**Pull request merged**